### PR TITLE
Eng 1134 elasticsearch

### PR DESCRIPTION
**Pull request by evonck**
Elasticsearch library.
The godoc link should work as soon as the PR is merged

### CODE REVIEW

**Comment by spierre** [comment]
tests pass for me

File: elasticsearch/elasticsearch.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+	return &val
+}
+
+// SearchUnique make a search in the Elastic index. using the Json query of the parameter.
+// Send back an error if it found no result or more than one.
+func (objectElk *Elastic) SearchUnique(query string) (string, error) {
+	val := objectElk.SearchElk(query)
+	if val == nil {
+		log.Warn("Nothing found")
+		return "", errors.New("No result ")
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: "No result " ?? "No result found" or "No result"

File: elasticsearch/elasticsearch.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+	}
+	if len(val.Hits.Hits) == 1 {
+		valueString, err := val.Hits.Hits[0].Source.MarshalJSON()
+		if err != nil {
+			log.Warn("Error while parsing the search result")
+			return "", errors.New("Error while parsing the search result")
+		}
+		return string(valueString), nil
+	} else if len(val.Hits.Hits) > 1 {
+		return "", errors.New("More than one result ")
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: why is there an ending space ?

File: elasticsearch/elasticsearch.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+		})
+		if err != nil {
+			return "", err
+		}
+		query := `{ "match": {"` + paramName + `": ` + string(jsonQueryURL) + `}}`
+		queries = append(queries, query)
+	}
+	elkQuery := `{"query" : {"bool" : {"must": [`
+	for i, queryString := range queries {
+		elkQuery += queryString
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: if i > 0 {
      elkQuery += ","
    }

Stop testing if last is such loop.

File: elasticsearch/elasticsearch.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+// CreateObject create a new object in elasticsearch.
+// Send back an error If an object with these ID already exist for the Elastic index
+// Create the index in ElasticSearch if needed
+func (objectElk *Elastic) CreateObject(id string, object interface{}) error {
+	connect := objectElk.GetConnection()
+	params := make(map[string]string)
+	params["id"] = id
+	value, err := objectElk.SearchUniqueParam(params)
+	if value != "" {
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: Stange code testing value assuming error is not empty.

File: elasticsearch/elasticsearch.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	connect.DeleteIndex(objectElk.IndexName)
+	connect.Close()
+}
+
+// CreateObject create a new object in elasticsearch.
+// Send back an error If an object with these ID already exist for the Elastic index
+// Create the index in ElasticSearch if needed
+func (objectElk *Elastic) CreateObject(id string, object interface{}) error {
+	connect := objectElk.GetConnection()
+	params := make(map[string]string)
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: this file is in go.
defer connect.close()

File: elasticsearch/elasticsearch.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+		log.Warn("Error while creating the object: ", err)
+	}
+	connect.Close()
+	return nil
+}
+
+// UpdateObject update or create a new object in the Elastic index
+// Create the index in ElasticSearch if needed
+func (objectElk *Elastic) UpdateObject(id string, object interface{}) {
+	connect := objectElk.GetConnection()
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: you are developing in GO.
defer connect.close()

File: elasticsearch/elasticsearch.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+	return objectElk.SearchUnique(elkQuery)
+}
+
+/***************************
+		Create/Update
+****************************/
+
+// ClearIndex clear the Elastic Index
+func (objectElk *Elastic) ClearIndex() {
+	connect := objectElk.GetConnection()
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: defer connect.close()

File: elasticsearch/elasticsearch.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+			continue
+		}
+		returnValue = append(returnValue, string(valueString))
+	}
+	return returnValue, nil
+}
+
+// SearchElk make a search in the Elastic index. using the Json query of the parameter
+func (objectElk *Elastic) SearchElk(query string) *elastigo.SearchResult {
+	connect := objectElk.GetConnection()
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: defer connect.close()

File: elasticsearch/elasticsearch.go [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+// Search make a search in the Elastic index. using the Json query of the parameter.
+// Send back an error if it found no result.
+func (objectElk *Elastic) Search(query string) ([]string, error) {
+	val := objectElk.SearchElk(query)
+	if val == nil {
+		log.Warn("Nothing found")
+		return nil, errors.New("No result ")
+	}
+	if len(val.Hits.Hits) == 0 {
+		log.Warn("Nothing found")
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: change this debug to empty result found. thats may help later.
en remode tailling space in error message.

File: elasticsearch/elasticsearch.go [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+		return nil, errors.New("No result ")
+	}
+	var returnValue []string
+	for i := range val.Hits.Hits {
+		valueString, err := val.Hits.Hits[i].Source.MarshalJSON()
+		if err != nil {
+			log.Warn("Error while parsing the search result")
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: Are you really thinking this debug may be usefull in the furur ?
add the *query* and the *i* index in the message, so it maight.

File: elasticsearch/elasticsearch.go [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+/***************************
+		Connection
+****************************/
+
+// GetConnection create the connection with ElasticSearch database
+func (objectElk *Elastic) GetConnection() *elastigo.Conn {
+	connect := elastigo.NewConn()
+	if objectElk.Host == "" {
+		log.Fatal("No connection with elasticsearch")
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: return nil !!

File: elasticsearch/elasticsearch.go [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+	return nil
+}
+
+// UpdateObject update or create a new object in the Elastic index
+// Create the index in ElasticSearch if needed
+func (objectElk *Elastic) UpdateObject(id string, object interface{}) {
+	connect := objectElk.GetConnection()
+	_, err := connect.Index(objectElk.IndexName, objectElk.IndexType, id, nil, object)
+	if err != nil {
+		log.Warn("Error while updating the object: ", err)
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: remove tailling space in "Error while updating the object: "

File: elasticsearch/README.md [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+Library elasticSearch containing the basic function used by meastor api micro service to communicate with [elasticsearch](https://www.elastic.co/products/elasticsearch).
+
+
+## Dependency
+-  [elastigo](https://github.com/mattbaird/elastigo) library
+
+## Run The Test
+In order to run the test of this library you need to have an elasticsearch server running.
+You can find a docker image of elasticsearch [here](https://hub.docker.com/_/elasticsearch/)
+
```
Comment (UrielCh) [inline elasticsearch/README.md]: add docker lines to start elastic search. I'm so lazy

File: elasticsearch/elasticsearch.go [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+			return "", err
+		}
+		query := `{ "match": {"` + paramName + `": ` + string(jsonQueryURL) + `}}`
+		queries = append(queries, query)
+	}
+	elkQuery := `{"query" : {"bool" : {"must": [`
+	for i, queryString := range queries {
+		if i > 0 {
+			elkQuery += ","
+
```
Comment (UrielCh) [inline elasticsearch/elasticsearch.go]: vide, pas bien !!

File: elasticsearch/elasticsearch_test.go [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+	convey.Convey("err should be nil", t, func() {
+		convey.So(err, convey.ShouldEqual, nil)
+	})
+	convey.Convey("project should be expectedStr0", t, func() {
+		convey.So(len(value), convey.ShouldEqual, len(expectedAll))
+	})
+}
+
+func TestSearch(t *testing.T) {
+	query := `{"query" : {"bool" : {"must": [{ "match": {"testParam": {"query":"testValue1","type":"phrase"}}}]}}}`
```
Comment (UrielCh) [inline elasticsearch/elasticsearch_test.go]: json With or without spacing ?

**Pull request merged**